static/js/unread_ops.js

```diff
@@ -9,32 +13,69 @@ import * as notifications from "./notifications";
 import * as people from "./people";
 import * as recent_topics_ui from "./recent_topics_ui";
 import * as recent_topics_util from "./recent_topics_util";
-import * as reload from "./reload";
+import * as ui_report from "./ui_report";
 import * as unread from "./unread";
 import * as unread_ui from "./unread_ui";
 
-export function mark_all_as_read() {
-    unread.declare_bankruptcy();
-    unread_ui.update_unread_counts();
+// anchor message is not included in this calculation
+const NUM_OF_MESSAGES_UPDATED_PER_BATCH = 5000;
+let loading_indicator_displayed = false;
+let unread_count_before_marking_all_msgs_read = 0;
+let messages_read_till_now = 0;
 
+export function mark_all_as_read() {
+    if (unread_count_before_marking_all_msgs_read === 0) {
+        unread_count_before_marking_all_msgs_read = unread.get_unread_message_count();
+    }
+    const opts = {
+        anchor: "first_unread",
```

Comment:
I think there was a thought that the recursive calls could use the anchor returned by the previous API response; this more obviously terminates and also is likely more efficient.

Reply:
Yeah, added that.